p2pserver/protocols/utils/cons_peers.go

```diff
@@ -0,0 +1,153 @@
+)
+
+type GovNodeResolver interface {
+	IsGovNode(key keypair.PublicKey) bool
+}
+
+type GovNodeMockResolver struct {
+	govNode map[string]struct{}
+}
+
+func NewGovNodeMockResolver(gov []string) *GovNodeMockResolver {
+	govNode := make(map[string]struct{}, len(gov))
+	for _, node := range gov {
+		govNode[node] = struct{}{}
+	}
+
+	return &GovNodeMockResolver{govNode}
+}
+
+func (self *GovNodeMockResolver) IsGovNode(key keypair.PublicKey) bool {
+	pubKey := vconfig.PubkeyID(key)
+	_, ok := self.govNode[pubKey]
+
+	return ok
+}
+
+type GovNodeLedgerResolver struct {
+	db   *ledger.Ledger
+	view uint32
+
+	cache unsafe.Pointer // atomic pointer to GovCache, avoid read&write data race
+}
+
+type GovCache struct {
+	pubkeys map[string]struct{}
+}
+
+func NewGovNodeResolver(db *ledger.Ledger) *GovNodeLedgerResolver {
+	return &GovNodeLedgerResolver{
+		db:    db,
+		cache: unsafe.Pointer(&GovCache{pubkeys: make(map[string]struct{})}),
+	}
+}
+
+func (self *GovNodeLedgerResolver) isGovNodeFromCache(pubkey string) bool {
+	cached := (*GovCache)(atomic.LoadPointer(&self.cache))
+	if cached != nil {
+		_, ok := cached.pubkeys[pubkey]
+		return ok
+	}
+
+	return false
+}
+
+func (self *GovNodeLedgerResolver) IsGovNode(key keypair.PublicKey) bool {
+	pubKey := vconfig.PubkeyID(key)
+	view, err := GetGovernanceView(self.db)
+	if err != nil || view.View == self.view {
+		return self.isGovNodeFromCache(pubKey)
+	}
+
+	govNode := false
+	peers, err := GetPeersConfig(self.db, view.View)
+	if err != nil {
+		return false
+	}
+
+	pubkeys := make(map[string]struct{}, len(peers))
+	for _, peer := range peers {
+		pubkeys[peer.PeerPubkey] = struct{}{}
+		if peer.PeerPubkey == pubKey {
+			govNode = true
+		}
+	}
+
+	atomic.StorePointer(&self.cache, unsafe.Pointer(&GovCache{pubkeys: pubkeys}))
+	self.view = view.View
+
+	return govNode
+}
+
+func GetGovernanceView(backend *ledger.Ledger) (*governance.GovernanceView, error) {
+	value, err := backend.GetStorageItem(utils.GovernanceContractAddress, []byte(governance.GOVERNANCE_VIEW))
+	if err != nil {
+		return nil, err
+	}
+	governanceView := new(governance.GovernanceView)
+	err = governanceView.Deserialize(bytes.NewBuffer(value))
+	if err != nil {
+		return nil, err
+	}
+	return governanceView, nil
+}
+
+func GetPeersConfig(backend *ledger.Ledger, view uint32) ([]*config.VBFTPeerStakeInfo, error) {
+	viewBytes := governance.GetUint32Bytes(view)
+	key := append([]byte(governance.PEER_POOL), viewBytes...)
+	data, err := backend.GetStorageItem(utils.GovernanceContractAddress, key)
+	if err != nil {
+		return nil, err
+	}
+	peerMap := &governance.PeerPoolMap{
+		PeerPoolMap: make(map[string]*governance.PeerPoolItem),
+	}
+	err = peerMap.Deserialization(common.NewZeroCopySource(data))
+	if err != nil {
+		return nil, err
+	}
+	var peerstakes []*config.VBFTPeerStakeInfo
+	for _, id := range peerMap.PeerPoolMap {
+		if id.Status == governance.CandidateStatus || id.Status == governance.ConsensusStatus {
```

Comment:
Should QuitCandidate & QuitConsensus be included either? @siovanus

Reply:
QuitCandidate means will disappare in next epoch, but exist in this epoch.

Reply:
if `QuitCandidate` is still candidate node, it can be in gov-subnet.